### PR TITLE
Concurrent sort example.

### PR DESCRIPTION
**Pull request by lucteo**
Using a mock for `spawn`/`await` so that the example runs without concurrent libraries.

### CODE REVIEW

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  precondition(n >= 8)
+  let stride = n / 8
+  let m1 = median3(a[0], a[stride], a[stride * 2])
+  let m2 = median3(a[stride * 3], a[stride * 4], a[stride * 5])
+  let m3 = median3(a[stride * 6], a[stride * 7], a[n - 1])
+  return median3(m1, m2, m3)
+}
+
+
+// TODO: this should be cleaned up and put in the standard library.
```
Comment (lucteo) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:64]: I would prefer this to be done in a separate PR.

File: StandardLibrary/Sources/Concurrency/Future.hylo [hunk 2]
```diff
@@ -46,6 +46,23 @@ public type EscapingFuture<E: Movable & Deinitializable>: Movable {
 
 }
 
+/// Mimics a future, but runs the computation inplace. Used for testing purposes.
```
Comment (kyouko-taiga) [inline StandardLibrary/Sources/Concurrency/Future.hylo:49]: If this type is only used for testing then it shouldn't belong to the standard library.

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    let (m1, m2) = sort_partition(&a)
+    inout (lhs, rhs) = &a.split(at: m1)
+    &rhs.drop_first(m2 - m1)
+
+    // Spawn work to sort the right-hand side.
+    let future = spawn_inplace(fun[sink let q=mutable_pointer[to: &rhs].copy()]() -> Int {
+      inout rhs = &(q.copy().unsafe[])
+      return my_concurrent_sort(&rhs)
+    })
+    // Execute the sorting on the left side, on the current thread.
```
Comment (kyouko-taiga) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:21]:
```suggestion

    // Execute the sorting on the left side, on the current thread.
```

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    let _ = my_concurrent_sort(&lhs)
+    let _ = future.await()
```
Comment (kyouko-taiga) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:23]:
```suggestion
    _ = my_concurrent_sort(&lhs)
    _ = future.await()
```
That's the canonical way to discard values.

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  }
+  return a.count()
+}
+
+fun sort_partition<Element: Regular & Comparable>(_ a: inout ArraySlice<Element>) -> {Int, Int} {
+  let mid_value = median9(a)
+  return a.partition(on: mid_value)
+}
+/// Returns the median of 3 values.
```
Comment (kyouko-taiga) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:32]:
```suggestion

/// Returns the median of 3 values.
```

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+      return my_concurrent_sort(&rhs)
+    })
+    // Execute the sorting on the left side, on the current thread.
+    let _ = my_concurrent_sort(&lhs)
+    let _ = future.await()
+  }
+  return a.count()
+}
+
+fun sort_partition<Element: Regular & Comparable>(_ a: inout ArraySlice<Element>) -> {Int, Int} {
```
Comment (kyouko-taiga) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:28]: Presumably this function is just computing a partition, right? If that's correct, it shouldn't be prefixed by `sort`.

We usually don't need to worry about polluting the global namespace in Hylo because non-public entities won't cross module boundaries. And if really we need namespacing, then we can use namespaces:
```
namespace ConcurrentSortExample {

  fun partition<Element: Regular & Comparable>(_ a: inout ArraySlice<Element>) -> {Int, Int} { ... }

}
```

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  let mid_value = median9(a)
+  return a.partition(on: mid_value)
+}
+/// Returns the median of 3 values.
+fun median3<Element: Regular & Comparable>(_ v1: Element, _ v2: Element, _ v3: Element) -> Element {
```
Comment (kyouko-taiga) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:33]:
```suggestion
fun median<Element: Regular & Comparable>(_ v1: Element, _ v2: Element, _ v3: Element) -> Element {
```
The fact that there are 3 values is encoded in the signature of the function.

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+      return v1.copy()
+    }
+    if v2 < v3 {
+      return v3.copy()
+    }
+    return v2.copy()
+  }
+}
+/// Returns a median of 9 values taken from the slice `a`.
+fun median9<Element: Regular & Comparable>(_ a: ArraySlice<Element>) -> Element {
```
Comment (kyouko-taiga) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:53]:
```suggestion

/// Returns a median of 9 values taken from the slice `a`.
fun median9<Element: Regular & Comparable>(_ a: ArraySlice<Element>) -> Element {
```
Again, I don't think the number of values should be in the stem of the function. I understand it can't be encoded in the structure of the function here (though I suspect we could eventually take a `Int[9]` that is projected out of the slice at the call site) so I would perhaps consider using an argument label instead.

Comment (kyouko-taiga) [reply Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:53]: I'm curious to know if @dabrahams has an opinion.

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  let n = a.count()
+  precondition(n >= 8)
+  let stride = n / 8
+  let m1 = median3(a[0], a[stride], a[stride * 2])
+  let m2 = median3(a[stride * 3], a[stride * 4], a[stride * 5])
+  let m3 = median3(a[stride * 6], a[stride * 7], a[n - 1])
+  return median3(m1, m2, m3)
+}
+
+
```
Comment (kyouko-taiga) [inline Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:63]:
```suggestion
```

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo (hunk 1 above)
Comment (kyouko-taiga) [reply Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:64]: Agreed. Dave has been working on a safe implementation but I think the support of remote values is not sufficient atm. I've started working on one that uses a `PointerToMutable`, like yours, but is generic over any kind of collection. I hope it will land very soon so you might want to wait a little bit.

If time is an issue you can always work with your implementation on a separate branch before we merge. I don't think it would be shocking even in a conference setting to say "all my examples can be ran on branch XYZ".

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo (hunk 4 above)
Comment (lucteo) [reply Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo]: Nice, I didn't know that. Thank you!

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo (hunk 8 above)
Comment (lucteo) [reply Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo:56]: Please note that this is different from the `median3` case. This one also selects the values used for computing the median from the given array slice. There is not indication in the type signature that it takes 9 values from the array slice.

File: StandardLibrary/Sources/Concurrency/Future.hylo (hunk 2 above)
Comment (lucteo) [reply StandardLibrary/Sources/Concurrency/Future.hylo:49]: Initially I had this in the example code, and I hesitate it whether to put it here or not.
I wanted this example to be also runnable by Dave, so I can move this back into the example code.

File: Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo (hunk 6 above)
Comment (lucteo) [reply Tests/EndToEndTests/TestCases/Concurrency/concurrent_sort.hylo]: Fixed.

**Review by kyouko-taiga — APPROVED** [review]
LGTM
